File: src/main/java/cspom/compiler/ConstraintParser.java
package cspom.compiler;

import java.util.LinkedList;
import java.util.List;

import scala.collection.JavaConversions;
import cspom.compiler.PredicateScanner.Node;
import cspom.constraint.FunctionalConstraint;
import cspom.constraint.GeneralConstraint;
import cspom.variable.CSPOMVariable;
import cspom.CSPOM;

public final class ConstraintParser {

    private final CSPOM problem;

    public ConstraintParser(final CSPOM problem) {
        this.problem = problem;
    }

    public void split(final String expression) throws PredicateParseException {
        final Node root = PredicateScanner.scan(expression);

        if (root.isLeaf()) {
            throw new IllegalArgumentException("Constraint expected");
        }
        final List<CSPOMVariable<?>> operands = new LinkedList<CSPOMVariable<?>>();
        for (Node n = root.getChild(); n != null; n = n.getSibling()) {
            operands.add(addToProblem(n));
        }
        problem.addConstraint(new GeneralConstraint(root.getOperator(), root
                .getParameters(), JavaConversions.asScalaBuffer(operands)));
    }

    private CSPOMVariable<?> addToProblem(final Node node)
            throws PredicateParseException {
        if (node.isLeaf()) {
            return addVariable(node, problem);
        }

        final CSPOMVariable<?> result = CSPOMVariable.aux();
        problem.addVariable(result);

        final List<CSPOMVariable<?>> operands = new LinkedList<CSPOMVariable<?>>();
        for (Node n = node.getChild(); n != null; n = n.getSibling()) {
            operands.add(addToProblem(n));
        }
        problem.addConstraint(new FunctionalConstraint(result, node
                .getOperator(), node.getParameters(), JavaConversions
                .asScalaBuffer(operands)));
        return result;
    }

    private static CSPOMVariable<?> addVariable(final Node node,
            final CSPOM problem) {
        final CSPOMVariable<?> existing = problem.variable(node.getOperator());
        if (existing != null) {
            return existing;
        }

        final CSPOMVariable<?> newVariable;
        if (node.isInteger()) {
            newVariable = CSPOMVariable.constant(Integer.parseInt(node
                    .getOperator()));
        } else {
            throw new IllegalStateException();
        }
        problem.addVariable(newVariable);
        return newVariable;
    }
}

File: src/main/scala/cspom/compiler/ConstraintParser.scala
package cspom.compiler;

import java.util.LinkedList
import java.util.List
import scala.collection.JavaConversions
import cspom.constraint.FunctionalConstraint
import cspom.constraint.GeneralConstraint
import cspom.variable.CSPOMVariable
import cspom.CSPOM
import scala.collection.mutable.Stack
import cspom.constraint.Predicate
import scala.collection.mutable.Stack
import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;
import java.util.regex.Pattern;

final object ConstraintParser {

  private val INTEGER = Pattern.compile("-?[0-9]*");

  private val IDENTIFIER = Pattern.compile("[a-zA-Z_]\\w*");

  def isInt(t: String) = INTEGER.matcher(t).matches

  def isId(t: String) = IDENTIFIER.matcher(t).matches

  def split(expression: String, problem: CSPOM) = {
    val root = scan(expression);

    require(!root.isLeaf, "Constraint expected");

    root.parameters.map { p =>
      problem.ctr(root.operator, p,
        root.child.toList.flatMap(_.siblings).map(addToProblem(_, problem)): _*)
    } getOrElse {
      problem.ctr(root.operator, root.child.toList.flatMap(_.siblings).map(addToProblem(_, problem)): _*)
    }
  }

  private def addToProblem(node: PredicateNode, problem: CSPOM): CSPOMVariable = {
    if (node.isLeaf) {
      problem.variable(node.operator).getOrElse {
        assume(node.isInteger, s"$node is not a valid leaf")
        problem.varOf(node.operator.toInt)
      }
    } else {
      node.parameters.map { p =>
        problem.is(node.operator, p, node.child.toList.flatMap(_.siblings).map(addToProblem(_, problem)): _*)
      } getOrElse {
        problem.is(node.operator, node.child.toList.flatMap(_.siblings).map(addToProblem(_, problem)): _*)
      }
    }
  }

  def scan(expression: String): PredicateNode = {
    val st = new StringTokenizer(expression, " {}(),", true);
    val stack = new Stack[PredicateNode]();
    var currentNode = new PredicateNode();
    var parameters: StringBuilder = null;

    while (st.hasMoreElements()) {
      val token = st.nextToken();
      if ("}" == token) {
        currentNode.parameters = Some(parameters.toString)
        parameters = null;
      } else if (parameters != null) {
        parameters.append(token);
      } else {
        token match {
          case " " =>
          case "{" =>
            require(currentNode.operatorIsDefined, "Empty operator");
            parameters = new StringBuilder();
          case "(" =>
            require(currentNode.operatorIsDefined, "Empty operator");
            val newPredicateNode = new PredicateNode();
            currentNode.child = Some(newPredicateNode)
            stack.push(currentNode);
            currentNode = newPredicateNode;
          case ")" =>
            require(!stack.isEmpty, "Too many )s");
            currentNode = stack.pop();
          case "," =>
            require(currentNode.operatorIsDefined, "Empty argument");
            val newPredicateNode = new PredicateNode();
            currentNode.sibling = Some(newPredicateNode)
            currentNode = newPredicateNode;
          case _ =>
            require(!currentNode.operatorIsDefined, "Delimiter expected in "
              + currentNode + " (" + expression + ")");
            currentNode.operator = token
        }
      }
    }
    currentNode;
  }

}
